src/java/htsjdk/samtools/sra/SRAUnalignmentIterator.java

```diff
@@ -0,0 +1,168 @@
+            readEnd = recordRangleInfo.getNumberOfReads();
+        } else if (readEnd <= 0) {
+            throw new RuntimeException("Invalid chunk provided: chunkStart position is after last read");
+        }
+
+        try {
+            unalignedIterator = run.getReadRange(readStart + 1, readEnd - readStart, Read.partiallyAligned | Read.unaligned);
+            nextUnalignedFragment();
+
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        // check unaligned
+        if (hasMoreUnalignedFragments == null) {
+            try {
+                lastRecord.detachFromIterator();
+                nextUnalignedFragment();
+            } catch (ErrorMsg e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return hasMoreUnalignedFragments;
+    }
+
+    @Override
+    public SAMRecord next() {
+        if (!hasNext()) {
+            return null;
+        }
+
+        return nextUnalignment();
+    }
+
+    @Override
+    public void remove() {
+        throw new RuntimeException("Removal of records not implemented.");
+    }
+
+    public void setValidationStringency(ValidationStringency validationStringency) {
+        this.validationStringency = validationStringency;
+    }
+
+    private SAMRecord nextUnalignment() {
+        try {
+            lastRecord = new SRALazyRecord(header, false, run, unalignedIterator, unalignedIterator.getReadId(), lastUnalignedFragmentIndex);
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+
+        if (validationStringency != null) {
+            lastRecord.setValidationStringency(validationStringency);
+        }
+
+        hasMoreUnalignedFragments = null;
+
+        return lastRecord;
+    }
+
+    private void nextUnalignedFragment() throws ErrorMsg {
+        while (hasMoreUnalignedFragments == null || hasMoreUnalignedFragments) {
+            hasMoreUnalignedFragments = unalignedIterator.nextFragment();
+            lastUnalignedFragmentIndex++;
+
+            if (hasMoreUnalignedFragments && !unalignedIterator.isAligned()) {
+                return;
+            }
+        }
+
+        if (!hasMoreUnalignedReads) {
+            throw new RuntimeException("Cannot get next unaligned read - already at last one");
+        }
+
+        while (true) {
+            hasMoreUnalignedReads = unalignedIterator.nextRead();
+            lastUnalignedFragmentIndex = -1;
+            if (!hasMoreUnalignedReads) {
+                break;
+            }
+            while (true) {
+                hasMoreUnalignedFragments = unalignedIterator.nextFragment();
+                lastUnalignedFragmentIndex++;
+                if (!hasMoreUnalignedFragments) {
+                    break;
+                }
+
+                if (!unalignedIterator.isAligned()) {
+                    break;
+                }
+            }
```

Comment:
Avoid `while (true)` when possible -- this one could be re-written as a `do....while(hasMoreUnalignedFragments && unalignedIterator.isAligned());` loop.